File: android/src/main/kotlin/com/gazepoint/sdk/PerformanceMonitor.kt
package com.gazepoint.sdk

import android.util.Log
import java.util.concurrent.ConcurrentLinkedQueue
import kotlin.math.roundToInt

/**
 * Performance monitor for tracking FPS, processing time, and other metrics.
 */
class PerformanceMonitor {

    companion object {
        private const val TAG = "PerformanceMonitor"
        private const val WINDOW_SIZE = 30
    }

    private val frameTimestamps = ConcurrentLinkedQueue<Long>()
    private val processingTimes = ConcurrentLinkedQueue<Long>()
    private var lastFrameTime = 0L
    private var frameCount = 0L
    private var totalProcessingTime = 0L

    data class PerformanceMetrics(
        val fps: Float,
        val avgProcessingTimeMs: Float,
        val maxProcessingTimeMs: Long,
        val droppedFrames: Int,
        val totalFrames: Long
    )

    fun startFrame(): Long = System.nanoTime()

    fun endFrame(startTime: Long) {
        val currentTime = System.nanoTime()
        val processingTime = (currentTime - startTime) / 1_000_000

        frameCount++
        totalProcessingTime += processingTime

        frameTimestamps.offer(currentTime)
        processingTimes.offer(processingTime)

        while (frameTimestamps.size > WINDOW_SIZE) {
            frameTimestamps.poll()
            processingTimes.poll()
        }

        lastFrameTime = currentTime
    }

    fun getMetrics(): PerformanceMetrics {
        if (frameTimestamps.isEmpty()) {
            return PerformanceMetrics(0f, 0f, 0L, 0, frameCount)
        }

        val timestamps = frameTimestamps.toList()
        val fps = if (timestamps.size > 1) {
            val timeSpan = (timestamps.last() - timestamps.first()) / 1_000_000_000.0
            if (timeSpan > 0) (timestamps.size / timeSpan).toFloat() else 0f
        } else {
            0f
        }

        val times = processingTimes.toList()
        val avgProcessingTime = if (times.isNotEmpty()) times.average().toFloat() else 0f
        val maxProcessingTime = times.maxOrNull() ?: 0L
        val targetFrameTime = 1000f / 30f
        val droppedFrames = times.count { it > targetFrameTime }

        return PerformanceMetrics(
            fps = fps,
            avgProcessingTimeMs = avgProcessingTime,
            maxProcessingTimeMs = maxProcessingTime,
            droppedFrames = droppedFrames,
            totalFrames = frameCount
        )
    }

    fun logMetrics() {
        val metrics = getMetrics()
        Log.d(
            TAG,
            """
            Performance Metrics:
            FPS: ${metrics.fps.roundToInt()}
            Avg Processing Time: ${metrics.avgProcessingTimeMs.roundToInt()} ms
            Max Processing Time: ${metrics.maxProcessingTimeMs} ms
            Dropped Frames: ${metrics.droppedFrames}
            Total Frames: ${metrics.totalFrames}
            """.trimIndent()
        )
    }

    fun reset() {
        frameTimestamps.clear()
        processingTimes.clear()
        lastFrameTime = 0L
        frameCount = 0L
        totalProcessingTime = 0L
    }

    fun isPerformanceDegraded(): Boolean {
        val metrics = getMetrics()
        return metrics.fps < 15f || metrics.avgProcessingTimeMs > 100f
    }
}

File: macos/gazepoint_sdk/Sources/gazepoint_sdk/GazePointSDK/PerformanceMonitor.swift
import Foundation

/// Performance monitor for tracking FPS, processing time, and other metrics
public class PerformanceMonitor {

    public struct PerformanceMetrics: Codable {
        public let fps: Float
        public let avgProcessingTimeMs: Float
        public let maxProcessingTimeMs: Float
        public let droppedFrames: Int
        public let totalFrames: Int64

        public init(
            fps: Float,
            avgProcessingTimeMs: Float,
            maxProcessingTimeMs: Float,
            droppedFrames: Int,
            totalFrames: Int64
        ) {
            self.fps = fps
            self.avgProcessingTimeMs = avgProcessingTimeMs
            self.maxProcessingTimeMs = maxProcessingTimeMs
            self.droppedFrames = droppedFrames
            self.totalFrames = totalFrames
        }
    }

    private let windowSize = 30
    private let targetFPS: Float = 30.0

    private var frameTimestamps: [TimeInterval] = []
    private var processingTimes: [TimeInterval] = []
    private var frameCount: Int64 = 0
    private var totalProcessingTime: TimeInterval = 0

    private let queue = DispatchQueue(
        label: "com.gazepoint.macos.performancemonitor",
        attributes: .concurrent
    )

    public init() {}

    public func startFrame() -> TimeInterval {
        return Date().timeIntervalSince1970
    }

    public func endFrame(startTime: TimeInterval) {
        let currentTime = Date().timeIntervalSince1970
        let processingTime = (currentTime - startTime) * 1000

        queue.async(flags: .barrier) { [weak self] in
            guard let self else { return }

            self.frameCount += 1
            self.totalProcessingTime += processingTime
            self.frameTimestamps.append(currentTime)
            self.processingTimes.append(processingTime)

            if self.frameTimestamps.count > self.windowSize {
                self.frameTimestamps.removeFirst()
                self.processingTimes.removeFirst()
            }
        }
    }

    public func getMetrics() -> PerformanceMetrics {
        return queue.sync {
            guard !frameTimestamps.isEmpty else {
                return PerformanceMetrics(
                    fps: 0,
                    avgProcessingTimeMs: 0,
                    maxProcessingTimeMs: 0,
                    droppedFrames: 0,
                    totalFrames: frameCount
                )
            }

            let fps: Float
            if frameTimestamps.count > 1 {
                let timeSpan = frameTimestamps.last! - frameTimestamps.first!
                fps = timeSpan > 0 ? Float(frameTimestamps.count) / Float(timeSpan) : 0
            } else {
                fps = 0
            }

            let avgProcessingTime = processingTimes.isEmpty
                ? 0
                : Float(processingTimes.reduce(0, +) / Double(processingTimes.count))
            let maxProcessingTime = processingTimes.max() ?? 0
            let targetFrameTime = 1000.0 / Double(targetFPS)
            let droppedFrames = processingTimes.filter { $0 > targetFrameTime }.count

            return PerformanceMetrics(
                fps: fps,
                avgProcessingTimeMs: avgProcessingTime,
                maxProcessingTimeMs: Float(maxProcessingTime),
                droppedFrames: droppedFrames,
                totalFrames: frameCount
            )
        }
    }

    public func reset() {
        queue.async(flags: .barrier) { [weak self] in
            self?.frameTimestamps.removeAll()
            self?.processingTimes.removeAll()
            self?.frameCount = 0
            self?.totalProcessingTime = 0
        }
    }
}
